internal/engine/compiler/impl_amd64.go

```diff
@@ -5030,3 +5100,45 @@ func (c *amd64Compiler) compileEnsureOnRegister(loc *runtimeValueLocation) (err
 	}
 	return
 }
+
+// compileMaybeSwapRegisters swaps two registers if they're not equal.
+func (c *amd64Compiler) compileMaybeSwapRegisters(reg1, reg2 asm.Register) {
+	if reg1 != reg2 {
+		c.assembler.CompileRegisterToRegister(amd64.XCHGQ, reg1, reg2)
+	}
+}
+
+// compilePreventCrossedTargetRegisters swaps registers in such a way, that for neither runtimeValueLocation from locs its
+// corresponding register with the same index from targets is occupied by some other runtimeValueLocation from locs. It returns a
+// closure to restore the original register placement.
+//
+// This function makes it possible to safely exchange one set of registers with another, where a register might be in both sets.
+// Each register will correspond either to itself or another register not present in its own set.
+//
+// For example, if we have locs = [AX, BX, CX], targets = [BX, SI, AX], then it'll do two swaps
+// to make locs = [BX, CX, AX]
+func (c *amd64Compiler) compilePreventCrossedTargetRegisters(locs []*runtimeValueLocation, targets []asm.Register) (restore func()) {
+	type Swap struct{ srcIndex, dstIndex int }
+	var swaps []Swap
+	for i := range locs {
+		targetLocation := -1 // -1 means not found.
+		for j := range locs {
+			if locs[j].register == targets[i] {
+				targetLocation = j
+				break
+			}
+		}
+		if targetLocation != -1 && targetLocation != i {
+			c.compileMaybeSwapRegisters(locs[i].register, locs[targetLocation].register)
+			locs[i].register, locs[targetLocation].register = locs[targetLocation].register, locs[i].register
+			swaps = append(swaps, Swap{i, targetLocation})
+		}
+	}
+	return func() {
+		for i := len(swaps) - 1; i >= 0; i -= 1 {
```

Comment:
why do we need to do reverse-order? trailing comment would be helpful for others.